README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

phoenix 将做为企业级基础组件:Scroll_anchor,提供项目支持。

## Features

phoenix 将做为企业级基础组件:Scroll_anchor,提供项目支持。

## Getting started

phoenix 将做为企业级基础组件:Scroll_anchor,提供项目支持。

## Usage

phoenix 将做为企业级基础组件:Scroll_anchor,提供项目支持。

```dart
const like = 'sample';
```

## Additional information

phoenix 将做为企业级基础组件:Scroll_anchor,提供项目支持。



CHANGELOG.md
<!--
 * @Author: lipeng 1162423147@qq.com
 * @Date: 2023-09-24 17:04:48
 * @LastEditors: lipeng 1162423147@qq.com
 * @LastEditTime: 2023-10-16 15:31:07
 * @FilePath: /phoenix_scroll_anchor/CHANGELOG.md
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
-->
## 1.0.0

* [新建] 企业级基础组件:Scroll_anchor


## 1.0.1

* [新增] example

## 1.0.2

* [修改] scroll_anchor_tab.dart

test/phoenix_scroll_anchor_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:phoenix_scroll_anchor/phoenix_scroll_anchor.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'scroll_actor_tab_example.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a blue toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: ScrollActorTabExample(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          //
          // TRY THIS: Invoke "debug painting" (choose the "Toggle Debug Paint"
          // action in the IDE, or press "p" in the console), to see the
          // wireframe for each widget.
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/scroll_anchor_tab.dart
import 'package:bindings_compatible/bindings_compatible.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:phoenix_popup/measure_size.dart';
import 'package:phoenix_tabbar/phoenix_tabbar.dart';

/// 构建指定索引的Widget
typedef AnchorTabWidgetIndexedBuilder = Widget Function(
    BuildContext context, int index);

/// 构建指定索引的Tab
typedef AnchorTabIndexedBuilder = BadgeTab Function(
    BuildContext context, int index);

/// 滑动锚点组件
class AnchorTab extends StatefulWidget {
  /// TabBar的样式
  final BrnAnchorTabBarStyle tabBarStyle;

  /// 构建指定索引的Widget
  final AnchorTabWidgetIndexedBuilder? widgetIndexedBuilder;

  /// 构建指定索引的Tab
  final AnchorTabIndexedBuilder tabIndexedBuilder;

  /// Tab与内容之间的分割线
  final Widget? tabDivider;

  /// 设置tab与widget的个数
  final int itemCount;

  AnchorTab(
      {required this.widgetIndexedBuilder,
      required this.tabIndexedBuilder,
      required this.itemCount,
      this.tabDivider,
      this.tabBarStyle = const BrnAnchorTabBarStyle()});

  @override
  _ScrollAnchorTabWidgetState createState() => _ScrollAnchorTabWidgetState();
}

class _ScrollAnchorTabWidgetState extends State<AnchorTab>
    with SingleTickerProviderStateMixin {
  /// 用于控制 滑动
  late ScrollController _scrollController;

  /// 用于控制tab
  late TabController _tabController;

  /// 滑动组件的 key
  late GlobalKey _key;

  /// 当前选中的索引
  int currentIndex = 0;

  /// 滑动组件的元素的key
  late List<GlobalKey> _bodyKeyList;

  /// 每个元素在滑动组件中的位置
  late List<double> _cardOffsetList;

  /// 是否点击滑动
  bool tab = false;

  /// 滑动组件在屏幕的位置
  double listDy = 0;

  @override
  void initState() {
    super.initState();
    _scrollController = ScrollController();

    _key = GlobalKey();
    _cardOffsetList = List.filled(widget.itemCount, -1.0);
    _bodyKeyList = [];
    currentIndex = 0;
    _tabController = TabController(length: widget.itemCount, vsync: this);

    fillKeyList();

    useWidgetsBinding().addPostFrameCallback((da) {
      fillOffset();
      _scrollController.addListener(() {
        _updateOffset();
        currentIndex = createIndex(_scrollController.offset);
        //防止再次 发送消息
        if (!tab) {
          _tabController.index = currentIndex;
        }
      });
    });
  }

  @override
  void didUpdateWidget(AnchorTab oldWidget) {
    super.didUpdateWidget(oldWidget);
    _cardOffsetList = List.filled(widget.itemCount, -1.0);
    int sub = widget.itemCount - oldWidget.itemCount;
    if (sub < 0) {
      _bodyKeyList = _bodyKeyList.sublist(0, widget.itemCount);
    } else if (sub > 0) {
      for (int i = 0; i < sub; i++) {
        _bodyKeyList.add(GlobalKey());
      }
    }
    if (sub != 0) {
      _tabController = TabController(length: widget.itemCount, vsync: this);
    }
    useWidgetsBinding().addPostFrameCallback((da) {
      fillOffset();
    });
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[
        PhoenixTabBar(
          indicatorColor: widget.tabBarStyle.indicatorColor,
          indicatorWeight: widget.tabBarStyle.indicatorWeight,
          indicatorPadding: widget.tabBarStyle.indicatorPadding,
          labelColor: widget.tabBarStyle.labelColor,
          labelStyle: widget.tabBarStyle.labelStyle,
          labelPadding: widget.tabBarStyle.labelPadding ?? EdgeInsets.zero,
          unselectedLabelColor: widget.tabBarStyle.unselectedLabelColor,
          unselectedLabelStyle: widget.tabBarStyle.unselectedLabelStyle,
          dragStartBehavior: widget.tabBarStyle.dragStartBehavior,
          controller: _tabController,
          tabs: _fillTab(),
          onTap: (state, index) {
            state.refreshBadgeState(index);
            currentIndex = index;
            tab = true;
            _scrollController
                .animateTo(_cardOffsetList[index],
                    duration: const Duration(milliseconds: 100),
                    curve: Curves.linear)
                .whenComplete(() {
              tab = false;
            });
          },
        ),
        widget.tabDivider ?? const SizedBox.shrink(),
        Expanded(
          child: SingleChildScrollView(
            key: _key,
            controller: _scrollController,
            child: Column(
              mainAxisSize: MainAxisSize.min,
              children: _fillList(),
            ),
          ),
        )
      ],
    );
  }

  List<Widget> _fillList() {
    List<Widget> tmpWidget = [];
    if (widget.widgetIndexedBuilder != null) {
      for (int i = 0, n = widget.itemCount; i < n; i++) {
        Widget itemWidget = Container(
            key: _bodyKeyList[i],
            child: widget.widgetIndexedBuilder!(context, i));
        itemWidget = MeasureSize(
          onChange: (size) {
            _updateOffset();
          },
          child: itemWidget,
        );
        tmpWidget.add(itemWidget);
      }
    }
    return tmpWidget;
  }

  void fillKeyList() {
    for (int i = 0, n = widget.itemCount; i < n; i++) {
      _bodyKeyList.add(GlobalKey());
    }
  }

  void fillOffset() {
    Offset globalToLocal =
        (_key.currentContext!.findRenderObject() as RenderBox)
            .localToGlobal(Offset.zero);
    listDy = globalToLocal.dy;

    for (int i = 0, n = widget.itemCount; i < n; i++) {
      if (_cardOffsetList[i] == -1.0 &&
          _bodyKeyList[i].currentContext != null) {
        double cardOffset =
            (_bodyKeyList[i].currentContext!.findRenderObject() as RenderBox)
                .localToGlobal(Offset.zero) //相对于原点 控件的位置
                .dy; //y点坐标

        _cardOffsetList[i] = cardOffset + _scrollController.offset - listDy;
      }
    }
  }

  List<BadgeTab> _fillTab() {
    List<BadgeTab> tmp = [];
    for (int i = 0, n = widget.itemCount; i < n; i++) {
      tmp.add(widget.tabIndexedBuilder(context, i));
    }
    return tmp;
  }

  void _updateOffset() {
    for (int i = 0, n = widget.itemCount; i < n; i++) {
      if (_bodyKeyList[i].currentContext != null) {
        double cardOffset =
            (_bodyKeyList[i].currentContext!.findRenderObject() as RenderBox)
                .localToGlobal(Offset.zero) //相对于原点 控件的位置
                .dy; //y点坐标

        _cardOffsetList[i] = cardOffset + _scrollController.offset - listDy;
      }
    }
  }

  /// 根据偏移量 确定tab索引
  int createIndex(double offset) {
    int index = widget.itemCount - 1;
    for (int i = 0; i < widget.itemCount - 1; i++) {
      if (offset >= _cardOffsetList[i] && (offset <= _cardOffsetList[i + 1])) {
        return i;
      }
    }
    return index;
  }

  @override
  void dispose() {
    super.dispose();
    _tabController.dispose();
    _scrollController.dispose();
  }
}

class BrnAnchorTabBarStyle {
  final Color? indicatorColor;

  final double indicatorWeight;

  final EdgeInsetsGeometry indicatorPadding;

  final Color? labelColor;

  final Color? unselectedLabelColor;

  final TextStyle? labelStyle;

  final EdgeInsetsGeometry? labelPadding;

  final TextStyle? unselectedLabelStyle;

  final DragStartBehavior dragStartBehavior;

  const BrnAnchorTabBarStyle({
    this.indicatorColor,
    this.indicatorWeight = 2.0,
    this.indicatorPadding = EdgeInsets.zero,
    this.labelColor,
    this.labelStyle,
    this.labelPadding,
    this.unselectedLabelColor,
    this.unselectedLabelStyle,
    this.dragStartBehavior = DragStartBehavior.start,
  });
}


example/lib/scroll_actor_tab_example.dart
/*
 * @Author: lipeng 1162423147@qq.com
 * @Date: 2022-04-29 17:06:50
 * @LastEditors: lipeng 1162423147@qq.com
 * @LastEditTime: 2023-10-12 23:19:55
 * @FilePath: /phoenix_scroll_anchor/example/lib/scroll_actor_tab_example.dart
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:phoenix_scroll_anchor/phoenix_scroll_anchor.dart';
import 'package:phoenix_tabbar/phoenix_tabbar.dart';

class ScrollActorTabExample extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('锚点'),
      ),
      body: AnchorTab(
        itemCount: 20,
        widgetIndexedBuilder: (context, index) {
          return StatefulBuilder(builder: (_, state) {
            double height = Random().nextInt(400).toDouble();
            return GestureDetector(
              child: Container(
                height: height,
                color: Color.fromARGB(
                    Random().nextInt(255),
                    Random().nextInt(255),
                    Random().nextInt(255),
                    Random().nextInt(255)),
                child: Center(child: Text('$index')),
              ),
              onTap: () {
                state(() {});
              },
            );
          });
        },
        tabIndexedBuilder: (context, index) {
          return BadgeTab(text: 'index $index');
        },
      ),
    );
  }
}
